### DISCUSSION TITLE
Nullify held key while still allowing it as modifier for combination

### DISCUSSION THREAD

**Discussion by pintassilgo** [xremap/xremap#315] [Q&A]
My mouse has 5 buttons, the two extras are to go back and forward, like in browser tab history. The command to go back or forward is triggered on press, not on release.

I take advantage of these buttons for combinations, like
- Hold forward button and turn mouse wheel to change audio volume.

If I use simply this:

```yaml
keymap:
  - remap:
      BTN_EXTRA-XUPSCROLL: KEY_VOLUMEUP
```

~It works,~ ¹ but the default command for BTN_EXTRA (navigate forward) is also triggered as soon as I hold the button. Navigate forward should only be triggered when I press the button alone, not in combination.
*¹: edit: actually it only works if I set `BTN_EXTRA` in `virtual_modifiers`, but then I'd lose default `BTN_EXTRA` command (navigate forward) and this is unwanted.*

The workaround I currently use is:

```yaml
modmap:
  - remap:
      BTN_EXTRA:
        held: KEY_UNKNOWN
        alone: BTN_EXTRA
        alone_timeout_millis: 250
        
virtual_modifiers:
  - KEY_UNKNOWN

keymap:
  - remap:
      KEY_UNKNOWN-XUPSCROLL: KEY_VOLUMEUP
      KEY_UNKNOWN-XDOWNSCROLL: KEY_VOLUMEDOWN
```

I don't know what `KEY_UNKNOWN` is, but as you can see I'm sacrificing it to be the null key for the combination. And this is my question: do I really need to sacrifice a key? There's no proper way to do it?

Because let's add a second remap: I want the combination "hold back button and turn mouse wheel" to be remapped to Ctrl+Tab (to switch active tab). So this is the code I'm using:

```yaml
modmap:
  - name: Side buttons only alone
    remap:
      BTN_EXTRA:
        held: KEY_UNKNOWN
        alone: BTN_EXTRA
        alone_timeout_millis: 250
      BTN_SIDE:
        held: BTN_9
        alone: BTN_SIDE
        alone_timeout_millis: 250
        
virtual_modifiers:
  - KEY_UNKNOWN
  - BTN_9

keymap:
  - name: Volumouse
    remap:
      KEY_UNKNOWN-XUPSCROLL: KEY_VOLUMEUP
      KEY_UNKNOWN-XDOWNSCROLL: KEY_VOLUMEDOWN

  - name: Switch tab
    remap:
      BTN_9-XUPSCROLL: Ctrl-Shift-Tab
      BTN_9-XDOWNSCROLL: Ctrl-Tab
```

I couldn't use `KEY_UNKNOWN` again, because it's already assigned. I had to sacrifice another "dumb" key, so I picked `BTN_9` that I have no clue what it's supposed to do.

It works, but this seems like a hack. I shouldn't need to discover keys that do nothing to sacrifice, like I did for `KEY_UNKNOWN` and `BTN_9`.

So I wonder if there's a better and proper way.

**Comment by k0kubun** [comment]
I think `KEY_RESERVED` could be used as another extra key instead of `KEY_9`.

The question is, what kind of config would you like to write if xremap had everything you need? Can you provide an example in YAML?

**Reply by pintassilgo** [reply]
> instead of `KEY_9`.

The key I'm sacrificing, actually a "button", is `BTN_9`, not `KEY_9`. Maybe there's a mouse with 9 buttons? I don't know what the 9th button of a mouse is supposed to do, so I guessed it was safe to discard it, so far it's fine.

> The question is, what kind of config would you like to write if xremap had everything you need? Can you provide an example in YAML?

Something like this:

```yaml
modmap:
  - name: Side buttons default command only alone
    remap:
      BTN_EXTRA:
        held: null
        alone: BTN_EXTRA
      BTN_SIDE:
        held: null
        alone: BTN_SIDE

keymap:
  - name: Volumouse
    remap:
      BTN_EXTRA-XUPSCROLL: KEY_VOLUMEUP
      BTN_EXTRA-XDOWNSCROLL: KEY_VOLUMEDOWN

  - name: Switch tab
    remap:
      BTN_SIDE-XUPSCROLL: Ctrl-Shift-Tab
      BTN_SIDE-XDOWNSCROLL: Ctrl-Tab
```

- First I make `BTN_EXTRA` and `BTN_SIDE` only run their default commands when pressed alone, doing nothing when held.
- The second part is similar to what I'm already using, but naming the buttons I'm actually pressing instead of using discarded keys like `KEY_UNKNOWN` and `BTN_9`.

In other words: a key set as `held: null` should automatically be registered as a `virtual_modifier` when held, while still being able to do something when pressed alone.

**Reply by pintassilgo** [reply]
Or, to be clearer, it could be `held: modifier` instead of `held: null`.

**Reply by k0kubun** [reply]
OK, thanks for the feedback. I plan to do some improvements in the internals soon, so I'll keep it in mind when I do that. In the meantime, I think your current workaround is the best I can think of.

**Comment by dlight** [comment]
> The workaround I currently use is:
>
> ```yaml
> modmap:
>   - remap:
>       BTN_EXTRA:
>         held: KEY_UNKNOWN
>         alone: BTN_EXTRA
>         alone_timeout_millis: 250
>         
> virtual_modifiers:
>   - KEY_UNKNOWN
> 
> keymap:
>   - remap:
>       KEY_UNKNOWN-XUPSCROLL: KEY_VOLUMEUP
>       KEY_UNKNOWN-XDOWNSCROLL: KEY_VOLUMEDOWN
> ```

Hey, thanks for this snippet! Maybe that's not the right place to ask, but: when I try this, the alone key (that is, sending a `BTN_EXTRA` when we press `BTN_EXTRA` alone) only trigger if I don't move the mouse while I'm clicking. Which is kind of difficult with a hi-dpi mouse. Is this a known issue?

**Reply by k0kubun** [reply]
Interesting. I can understand why this happens; `held` key is emitted when something else is pressed, and a mouse move is considered a "press" when xremap is attached to a mouse as well. I think we could let xremap avoid emitting `held` when such events are detected. I'd be open to merging a PR that implements it. But at the moment, it's how `alone`/`held` behaves.

**Reply by dlight** [reply]
But is it useful to have mouse motion being handled as a key, at all?

Should mouse movements always be dropped, or only ignored when there is a modifier being held?

My reasoning is that if it is ever useful to trigger on a mouse movement, you probably want to trigger on a modifier plus a mouse movement as well. But I think that this basically never happens, and nobody is using xremap to remap mouse movements (what does this mean exactly?? maybe pressing a button plus movement being used to control volume?)

Also: what key or keys represent a mouse movement? (how to find out? I'm using something like `cargo run --example evtest` from [evdev](https://github.com/emberian/evdev) but it doesn't output an identifier from [here](https://github.com/emberian/evdev/blob/main/src/scancodes.rs#L26-L572) on a mouse movement)

**Reply by k0kubun** [reply]
> But is it useful to have mouse motion being handled as a key, at all?

These events are not supposed to impact any remapping behaviors, and I guess it effectively works that way in most cases, but you hit a corner case.

> how to find out?

I think `RUST_LOG=debug xremap ...` would print the events when you move a mouse. If it doesn't, I guess we need to add logging for mouse events.
